### DISCUSSION TITLE
How would I programatically create an Astro subtree?

### DISCUSSION THREAD

**Discussion by flexdinesh** [withastro/roadmap#378] [Proposal]
I've been poking around [markdoc](https://markdoc.dev/) to get it to work with Astro. Markdoc parses markdown to an AST tree and will allow me to use custom non-markdown elements in markdown files with type safety. It's my preferred tool over MDX.

Markdoc allows us to build our own renderer and I'm keen to build one for Astro. Eg. [Markdoc React renderer](https://github.com/markdoc/markdoc/blob/main/src/renderers/react/react.ts). To do so, I want to be able to create an Astro subtree programatically in a `.ts` file. Is there an example somewhere I could refer to create Astro nodes using `Astro` API like `React.createElement()`?

It is important to be able to create an Astro subtree in a `.ts` file because I would need to recursively invoke the same method. 

I got things to work a bit but it's not very pretty and I can't publish this as a markdoc renderer package since the tree is constructed in a `.astro` file. This is my current approach.

```tsx
// MarkdownRenderer.astro
---
import type { RenderableTreeNodes } from "@markdoc/markdoc";
// I want to be able to render the tree within this .ts file
import { Node } from "./node-helper";

type Props = {
  content: RenderableTreeNodes;
  components: Record<
    string,
    {
      Component: any; // ASTRO component
      props: Record<string, string | number>;
    }
  >;
};

const { content, components } = Astro.props;

const node = new Node(content, components);
node.validateElement();

const Tag = node.tag;
const props = node.props;
const children = node.children;

---

{node.shouldRenderChildren() ? children : null}

{
  node.shouldRenderSelf() ? 
      content.map((element) => (
        <Astro.self content={element} components={components} />
      ))
    : null
}

{
  node.shouldRenderTag() ? (
    <Tag {...props}>
      {node.hasChildren() ? (
        <Astro.self content={children} components={components} />
      ) : null}
    </Tag>
  ) : null
}

```

Instead of this I would like to be able to render the whole tree in a ts file and just invoke it in an `.astro` file like this

```tsx
// MarkdownRenderer.astro
---
...
const { content, components } = Astro.props;

const node = new Node(content, components);
node.validateElement();


---
{node.renderTree()}

```

**Comment by natemoo-re** [comment]
Thanks for asking about this! We do have a programmatic JavaScript API, but it's not documented at the moment since we've been considering it internal and unstable. It's mostly used for internal unit tests at the moment. 

But this suggestion is making me think that we should definitely consider exposing a stable public API for it!

**Reply by natemoo-re** [reply]
Here is a [sample unit test](https://github.com/withastro/astro/blob/main/packages/astro/test/units/render/jsx.test.js#L33-L35) where the API is currently used manually.

**Reply by flexdinesh** [reply]
I have currently published the markdoc renderer as an Astro component to npm.
https://www.npmjs.com/package/astro-markdoc-renderer

It works but something about having the renderer as an Astro component doesn't feel right. I'll go through the tests and see if I could convert the markdoc renderer into a javascript package. Thanks for sharing the link to a code example @natemoo-re.

**Comment by florian-lefebvre** [comment]
Markdoc is now officially supported: https://docs.astro.build/en/guides/integrations-guide/markdoc/
